internal/defact/default.go

```diff
@@ -0,0 +1,262 @@
+
+	// remove symlinks pointing to default project
+	files, err := ioutil.ReadDir(binDir)
+	if err != nil {
+		return err
+	}
+	for _, f := range files {
+		fn := filepath.Join(binDir, f.Name())
+		if !isShimAndTargetsDir(fn, config.CachePath()) {
+			continue
+		}
+
+		// remove shim if it links to old project path or target does not exist anymore
+		err = os.Remove(fn)
+		if err != nil {
+			return locale.WrapError(err, "rollback_remove_err", "Failed to remove shim {{.V0}}", fn)
+		}
+	}
+	return nil
+}
+
+// Executables will return all the Executables that need to be symlinked in the various provided bin directories
+func Executables(bins []string) ([]string, error) {
+	exes := []string{}
+
+	for _, bin := range bins {
+		err := filepath.Walk(bin, func(fpath string, info os.FileInfo, err error) error {
+			// Filter out files that are not executable
+			if info == nil || info.IsDir() || !fileutils.IsExecutable(fpath) { // check if executable by anyone
+				return nil // not executable
+			}
+
+			exes = append(exes, fpath)
+			return nil
+		})
+		if err != nil {
+			return exes, errs.Wrap(err, "Error while walking path")
+		}
+	}
+
+	return exes, nil
+}
+
+// UniqueExes filters the array of executables for those that would be selected by the command shell in case of a name collision
+func UniqueExes(exePaths []string, pathext string) ([]string, error) {
+	pathExt := strings.Split(strings.ToLower(pathext), ";")
+	exeFiles := map[string]exeFile{}
+	result := []string{}
+
+	for _, exePath := range exePaths {
+		if rt.GOOS == "windows" {
+			exePath = strings.ToLower(exePath) // Windows is case-insensitive
+		}
+
+		exe := exeFile{exePath, "", ""}
+		ext := filepath.Ext(exePath)
+
+		// We only set the executable extension if PATHEXT is present.
+		// Some macOS builds can contain binaries with periods in their
+		// names and we do not want to strip off suffixes after the period.
+		if funk.Contains(pathExt, ext) {
+			exe.ext = filepath.Ext(exePath)
+		}
+		exe.name = strings.TrimSuffix(filepath.Base(exePath), exe.ext)
+
+		if prevExe, exists := exeFiles[exe.name]; exists {
+			pathsEqual, err := fileutils.PathsEqual(filepath.Dir(exe.fpath), filepath.Dir(prevExe.fpath))
+			if err != nil {
+				return result, errs.Wrap(err, "Could not compare paths")
+			}
+			if !pathsEqual {
+				continue // Earlier PATH entries win
+			}
+			if funk.IndexOf(pathExt, prevExe.ext) < funk.IndexOf(pathExt, exe.ext) {
+				continue // Earlier PATHEXT entries win
+			}
+		}
+
+		exeFiles[exe.name] = exe
+	}
+
+	for _, exe := range exeFiles {
+		result = append(result, exe.fpath)
+	}
+	return result, nil
+}
+
+// shimTargetPath returns the full path to the shim target (adds .bat on Windows)
+func shimTargetPath(targetDir string, path string) string {
+	target := filepath.Clean(filepath.Join(targetDir, filepath.Base(path)))
+	if rt.GOOS != "windows" {
+		return target
+	}
+
+	oldExt := filepath.Ext(target)
+	return target[0:len(target)-len(oldExt)] + ".bat"
```

Comment:
scriptfile and the language type should be the ones facilitating this

http://github.com/ActiveState/cli/blob/pkg-174895290/internal/scriptfile/scriptfile.go#L20-L20